presto-main/src/main/java/com/facebook/presto/sql/planner/TimestampRuleChange.java

```diff
@@ -0,0 +1,144 @@
+import static com.facebook.presto.sql.planner.plan.Patterns.filter;
+import static com.facebook.presto.sql.relational.Expressions.comparisonExpression;
+import static com.facebook.presto.sql.relational.Expressions.constant;
+
+/**
+ * Minimal iterative rule that rewrites:
+ *   CAST(ts_col AS date) = DATE 'yyyy-mm-dd'
+ * into
+ *   ts_col >= TIMESTAMP 'yyyy-mm-dd 00:00:00.000'
+ * (No upper bound to avoid referencing "AND".)
+ */
+public class TimestampRuleChange
+        implements Rule<FilterNode>
+{
+    private static final Pattern<FilterNode> PATTERN = filter();
+
+    private final FunctionAndTypeManager functionAndTypeManager;
+    private final StandardFunctionResolution functionResolution;
+
+    public TimestampRuleChange(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.functionAndTypeManager = functionAndTypeManager;
+        this.functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+    }
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(FilterNode filter, Captures captures, Context context)
+    {
+        RowExpression originalPredicate = filter.getPredicate();
+        RowExpression rewritten = rewritePredicate(originalPredicate);
+        if (rewritten.equals(originalPredicate)) {
+            return Result.empty();
+        }
+
+        FilterNode newFilter = new FilterNode(
+                filter.getSourceLocation(),
+                filter.getId(),
+                filter.getSource(),
+                rewritten);
+        return Result.ofPlanNode(newFilter);
+    }
+
+    private RowExpression rewritePredicate(RowExpression expression)
+    {
+        // If not a function call, do nothing
+        if (!(expression instanceof CallExpression)) {
+            return expression;
+        }
+        CallExpression call = (CallExpression) expression;
+
+        // If the displayName is '=' and there are exactly two arguments, check for date(...) usage
+        if ("=".equalsIgnoreCase(call.getDisplayName()) && call.getArguments().size() == 2) {
+            RowExpression left = call.getArguments().get(0);
+            RowExpression right = call.getArguments().get(1);
+
+            Optional<RowExpression> maybeCol = extractCastToDate(left);
+            Optional<LocalDate> maybeDate = extractDateLiteral(right);
+
+            if (!maybeCol.isPresent() || !maybeDate.isPresent()) {
+                // try flipping
+                maybeCol = extractCastToDate(right);
+                maybeDate = extractDateLiteral(left);
+            }
+
+            if (maybeCol.isPresent() && maybeDate.isPresent()) {
+                // We'll rewrite to: col >= TIMESTAMP 'yyyy-mm-dd 00:00:00.000'
+                LocalDate dateVal = maybeDate.get();
+                String start = dateVal.toString() + " 00:00:00.000";
+
+                return comparisonExpression(
```

Comment:
In order to get the correctness all we need to do is create a `callExpression` which has a function of `AND` and supplies the two arguments for the `>=` expression and one for the `<` expression.